### PR TITLE
Slot state transition with virtual skip block

### PR DESCRIPTION
**Pull request by djrtwo**
A modified version of #234 that uses a "virtual" skip block instead of inserting an actual skip block into the chain. This proposal removes the "opaque-ness" of dealing with previously unknown hashes of skip blocks.

There might be minor bugs that I'll fix after some initial review/thought if we decide to go with this proposal

### CODE REVIEW

**Comment by hwwhww** [comment]
So the skip block only has two meaningful fields:
1. `slot`
2. `ancestor_hashes`

If `skip_block is True`:
* Don’t process
	1. Proposer signature
	2. Attestations
	3. PoW receipt root
	4. Special objects 

* Process
	2. Some RANDAO
	3. *Most* per-epoch processing
	4. Verify block.state_root (state root processing, post-state-transition)

For readability, maybe in the per-slot transaction chapter, we can add a section “processing skip block”  that contains all description of  "the procedure of processing skip slot per-slot", and then we can jump to per-epoch chapter earlier.

**Comment by JustinDrake** [comment]
I think no skip blocks are the way forward 👍. (Technically the two approaches are probably equivalent because `previous_proposal_hash` can be provided at the network layer and the client can make it progress with an appropriate number of skip blocks to check it's consistent with `previous_block_hash`.)

Because there's no more `BeaconBlock` object I would call it a skip *slot* (as opposed to a skip *block*).

Feel free to close https://github.com/ethereum/eth2.0-specs/pull/234. (@hwwhww has made comments there which could be cherry-picked.)

File: specs/core/0_beacon-chain.md [hunk 1]
```diff
@@ -1592,9 +1587,13 @@ while len(state.persistent_committee_reassignments) > 0 and state.persistent_com
 ### Final updates
 
 * Remove any `attestation` in `state.latest_attestations` such that `attestation.data.slot < s`.
-* Run `exit_validator(i, state, penalize=False, current_slot=block.slot)` for indices `i` such that `state.validator_registry[i].status == ACTIVE and state.validator_registry[i].balance < MIN_BALANCE`.
+* Run `exit_validator(i, state, penalize=False, current_slot=state.slot)` for indices `i` such that `state.validator_registry[i].status == ACTIVE and state.validator_registry[i].balance < MIN_BALANCE`.
 * Set `state.latest_block_hashes = state.latest_block_hashes[EPOCH_LENGTH:]`.
-* Set `state.latest_state_recalculation_slot += EPOCH_LENGTH`.
+
+## State root processing
```
[the remark this answers is not in the document]
Comment (djrtwo) [reply specs/core/0_beacon-chain.md]: Can't check the state root until the entirety of the state has been updated during a particular transition.

Has to be at the end of everything. Might be able to swap position of per-epoch with per-slot and put this at the end of per-slot in that case.

I need to think about the implications of the ordering there and leave it to a separate PR

File: specs/core/0_beacon-chain.md [hunk 2]
```diff
@@ -1338,15 +1336,13 @@ process_deposit(
...
 )
 ```
 
-## Epoch boundary processing
+## Per-epoch processing
 
-Repeat the steps in this section while `block.slot - state.latest_state_recalculation_slot >= EPOCH_LENGTH`. For simplicity, we use `s` as `state.latest_state_recalculation_slot`.
-
-Note that `state.latest_state_recalculation_slot` will always be a multiple of `EPOCH_LENGTH`. In the "happy case", this process will trigger, and loop once, every time `block.slot` passes a new exact multiple of `EPOCH_LENGTH`, but if a chain skips more than an entire epoch then the loop may run multiple times, incrementing `state.latest_state_recalculation_slot` by `EPOCH_LENGTH` with each iteration.
+The steps below happen when `state.slot % EPOCH_LENGTH == 0`. For simplicity we denote `state.slot - EPOCH_LENGTH` by `s`.
```
[the remark this answers is not in the document]
Comment (djrtwo) [reply specs/core/0_beacon-chain.md]: very much agreed. I'll take a stab at it

File: specs/core/0_beacon-chain.md [hunk 3]
```diff
@@ -1355,15 +1355,15 @@ All [validators](#dfn-validator):
...
+* Let `this_epoch_boundary_attestations = [a for a in this_epoch_attestations if a.data.epoch_boundary_hash == get_block_hash(state, state.slot-EPOCH_LENGTH) and a.justified_slot == state.justified_slot]`.
 * Let `this_epoch_boundary_attesters` be the union of the [validator](#dfn-validator) index sets given by `[get_attestation_participants(state, a.data, a.participation_bitfield) for a in this_epoch_boundary_attestations]`.
 * Let `this_epoch_boundary_attesting_balance = sum([get_effective_balance(v) for v in this_epoch_boundary_attesters])`.
 
 [Validators](#dfn-Validator) justifying the epoch boundary block at the start of the previous epoch:
 
-* Let `previous_epoch_attestations = [a for a in state.latest_attestations if s - EPOCH_LENGTH <= a.slot < s]`.
-* Let `previous_epoch_boundary_attestations = [a for a in this_epoch_attestations + previous_epoch_attestations if a.epoch_boundary_hash == get_block_hash(state, s - EPOCH_LENGTH) and a.justified_slot == state.previous_justified_slot]`.
+* Let `previous_epoch_attestations = [a for a in state.latest_attestations if state.slot - 2*EPOCH_LENGTH <= a.slot < state.slot - EPOCH_LENGTH]`.
```
Comment (JustinDrake) [inline specs/core/0_beacon-chain.md]: Be consistent with spacing, e.g. `2*EPOCH_LENGTH` vs `2 * EPOCH_LENGTH`.

**Pull request merged**